core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -282,19 +282,23 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
     //
     // if valid, transition is done via overwriting the whole object to ensure synchronization
 
-    val toState = pendingState.getOrElse(throw new IllegalStateException(s"TransactionalId $transactionalId " +
-      "completing transaction state transition while it does not have a pending state"))
+    val toState = pendingState.getOrElse {
+      fatal(s"${this.toString}'s transition to $toState failed since pendingState is not defined: this should not happen")
```

Comment:
nit: The `toString` is redundant, no? Could this just be `error`? We don't actually do anything if it happens.